### PR TITLE
Support compress for RISC-V

### PR DESCRIPTION
**Pull request by ita-sc**
Hi

This patch adds support for `compress` for RISC-V.

Also I've added some needed support for `store`, `count_true` and minor support for `convert` implemented. For now unit tests for `convert` can not be added, as we have some problems with saturated arithmetics here.

### CODE REVIEW

File: include/eve/module/core/regular/impl/simd/riscv/store.hpp [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+EVE_FORCEINLINE void
+store_(EVE_SUPPORTS(rvv_), C const& cond, wide<T, N> const& v, Ptr ptr) noexcept
+requires rvv_abi<abi_t<T, N>> && (!has_store_equivalent<wide<T, N>, Ptr>)
+{
+  if constexpr( C::is_complete || C::has_alternative || N() != expected_cardinal_v<T> )
+  {
+    store_(EVE_RETARGET(cpu_), cond, v, ptr);
+  }
+  else riscv_store(cond.mask(as<wide<T, N>> {}), v, unalign(ptr));
+}
```
Comment (DenisYaroshevskiy) [inline include/eve/module/core/regular/impl/simd/riscv/store.hpp:68]: N() != expected cardinal strikes me as weird. expected cardinal is enormous. why can't you do a masked store for small register

File: test/unit/api/compress/compress_copy_test.hpp [hunk 2]
```diff
@@ -9,6 +9,29 @@
 
 #include "test.hpp"
 
+template<typename Type> struct rvv_compress_filter
```
Comment (DenisYaroshevskiy) [inline test/unit/api/compress/compress_copy_test.hpp:12]: Why is rvv compress filter different from any other rvv filter. This doesn't seem right

File: test/unit/api/compress/compress_copy_test.hpp [hunk 3]
```diff
@@ -151,7 +174,8 @@ compress_copy_tst_mask(eve::as<T> tgt, L m, auto algo)
       for( int j = 0; j != T::size() - i; ++j ) { for_ignore_tests(eve::ignore_extrema(i, j)); }
     }
   }
-  else if constexpr( eve::has_bundle_abi_v<T> && eve::current_api >= eve::sve )
+  else if constexpr( eve::has_bundle_abi_v<T> && eve::current_api >= eve::sve
+                     && eve::current_api >= eve::rvv )
```
Comment (DenisYaroshevskiy) [inline test/unit/api/compress/compress_copy_test.hpp:178]: && doesn't seem right. this is always false.

Comment (ita-sc) [reply test/unit/api/compress/compress_copy_test.hpp]: Yep, fixed.

File: test/unit/api/compress/compress_copy_test.hpp (hunk 2 above)
Comment (ita-sc) [reply test/unit/api/compress/compress_copy_test.hpp:12]: Actually, this is the main reason why I start with compress  (and next patch will be about shuffles). They require a lot of memory to compile even with global reduction for testcases. So, for now I've created an additional filter for these tests, though it is still not clear how to make it better.

File: include/eve/module/core/regular/impl/simd/riscv/store.hpp (hunk 1 above)
Comment (ita-sc) [reply include/eve/module/core/regular/impl/simd/riscv/store.hpp:75]: Yep, removed. Also tried to make some optimizations for store. Take a look please.
Note: I have some doubts that the case of `C::is_complete && !C::is_inverted` is good (I just do nothing in this case), though it seems like the tests pass.

File: test/unit/api/compress/compress_wide.cpp [hunk 4]
```diff
@@ -1,9 +1,12 @@
...
 
-TTS_CASE_WITH( "Check compress behavior"
-        , eve::test::simd::all_types
-        , tts::generate(tts::ramp(1))
-        )
+#ifdef SPY_SIMD_IS_RISCV_FIXED_RVV
+using types = tts::filter<rvv_compress_filter, eve::test::simd::all_types>::type;
+#else
+using types = eve::test::simd::all_types;
+#endif
```
Comment (DenisYaroshevskiy) [inline test/unit/api/compress/compress_wide.cpp:7]: I'd suggest: 

put this in comoress tests

call it simd_types_for_compress

Comment (ita-sc) [reply test/unit/api/compress/compress_wide.cpp]: Done

**Comment by ita-sc** [comment]
Gentle ping

File: include/eve/module/core/compress/simd/riscv/compress.hpp [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+
+template<relative_conditional_expr C, typename T, typename N, typename U>
+EVE_FORCEINLINE auto
+compress_(EVE_SUPPORTS(rvv_), C c, wide<T, N> v, logical<wide<U, N>> mask) noexcept
+requires rvv_abi<abi_t<T, N>>
+{
+  auto                c_mask          = expand_mask(c, as<wide<T, N>> {});
+  auto                formated_mask   = simd_cast(mask, as<logical<wide<T, N>>> {});
+  logical<wide<T, N>> real_mask       = __riscv_vmand(c_mask, formated_mask, N::value);
+  wide<T, N>          init            = make(as<wide<T, N>> {}, static_cast<T>(0));
```
Comment (DenisYaroshevskiy) [inline include/eve/module/core/compress/simd/riscv/compress.hpp:21]: `wide<T, N> init(0)` or `eve::zero(eve::as(v))`

File: include/eve/module/core/regular/impl/simd/riscv/count_true.hpp [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+
+#include <eve/concept/conditional.hpp>
+#include <eve/concept/value.hpp>
+#include <eve/detail/implementation.hpp>
+
+namespace eve::detail
+{
+template<scalar_value T, typename N, relative_conditional_expr C>
+EVE_FORCEINLINE std::ptrdiff_t
+                count_true_(EVE_SUPPORTS(cpu_), C cond, logical<wide<T, N>> v) noexcept
```
Comment (DenisYaroshevskiy) [inline include/eve/module/core/regular/impl/simd/riscv/count_true.hpp:18]: You should check for `C::is_complete` and not call the masked versions for those.

(is_complete is ignore_all, ignore_none)
ignore_all - is_inverted == false
ignore_none - is_inverted == true

File: include/eve/module/core/regular/impl/simd/riscv/count_true.hpp [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+{
+  return __riscv_vcpop(v, N::value);
+}
+template<scalar_value T, typename N>
+EVE_FORCEINLINE std::ptrdiff_t
+                count_true_(EVE_SUPPORTS(rvv_), top_bits<logical<wide<T, N>>> v) noexcept
+requires rvv_abi<abi_t<T, N>>
+{
+  return __riscv_vcpop(v.storage, N::value);
+}
```
Comment (DenisYaroshevskiy) [inline include/eve/module/core/regular/impl/simd/riscv/count_true.hpp:38]: I'd prefere if for both of these you'd call the ignore version and handle masked not masked there.

File: include/eve/module/core/regular/impl/simd/riscv/store.hpp (hunk 1 above)
Comment (DenisYaroshevskiy) [reply include/eve/module/core/regular/impl/simd/riscv/store.hpp:75]: Yeah - that terminology is weird. !C::is_inverted means everything is ignored.

File: include/eve/module/core/regular/impl/simd/riscv/store.hpp [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+         relative_conditional_expr       C,
+         simd_compatible_ptr<wide<T, N>> Ptr>
+EVE_FORCEINLINE void
+store_(EVE_SUPPORTS(rvv_), C const& cond, wide<T, N> const& v, Ptr ptr) noexcept
+requires rvv_abi<abi_t<T, N>> && (!has_store_equivalent<wide<T, N>, Ptr>)
+{
+  auto p = unalign(ptr);
+  if constexpr( C::is_complete )
+  {
+    if constexpr( !C::is_inverted ) return;
```
Comment (DenisYaroshevskiy) [inline include/eve/module/core/regular/impl/simd/riscv/store.hpp:66]: !is_inverted but has alternative should write the thing.

https://github.com/jfalcou/eve/blob/c2dacd782c29177ffc4b8ef86a900d6bc6389e67/include/eve/module/core/regular/impl/store.hpp#L119

Did you enable the tests for store?

File: test/unit/api/compress/compress_test_filter.hpp [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+    constexpr auto cardinal          = eve::cardinal_v<Type>;
+    constexpr auto expected_cardinal = eve::expected_cardinal_v<scalar_t>;
+    // and double
+    if( sizeof(scalar_t) != 8 || !std::is_floating_point_v<scalar_t> ) return false;
+    // for expected cardinal (lmul==8)
+    if( expected_cardinal == cardinal ) return true;
+    // for combined type
+    if( 2 * expected_cardinal == cardinal ) return true;
+    return false;
+  }
```
Comment (DenisYaroshevskiy) [inline test/unit/api/compress/compress_test_filter.hpp:28]: I find this function way too difficult to parse through. What's bit_size / 8.

sizeof(Type) <= bit_size / 8 should be true for all types. What are you trying to do. 

If you explain in plain english what you want, we can do figure out how to write it.

**Review by DenisYaroshevskiy — COMMENTED** [review]
Sorry for slow responses.

Most of this is ~fine - some small things.

The reduction in test cases is where we have most problems.

File: include/eve/module/core/regular/impl/simd/riscv/store.hpp (hunk 1 above)
Comment (jfalcou) [reply include/eve/module/core/regular/impl/simd/riscv/store.hpp:75]: we shoudl probably have proper concept for that instead of those mysterious naming scheme

File: .github/workflows/unit.yml [hunk 10]
```diff
@@ -147,9 +147,9 @@ jobs:
           mkdir build && cd build
           cmake .. -G Ninja -DEVE_OPTIONS="${{ matrix.cfg.opts }}" -DCMAKE_TOOLCHAIN_FILE=../cmake/toolchain/${{ matrix.cfg.comp }}.${{ matrix.cfg.arch }}.cmake
       - name: Compiling Unit Tests
-        run:  cd build && ninja unit.meta.exe unit.arch.exe  unit.core.bit_cast.exe unit.api.regular.wide.exe unit.core.simd_cast.exe -j 5
+        run:  cd build && ninja unit.meta.exe unit.arch.exe  unit.core.bit_cast.exe unit.api.regular.wide.exe unit.core.simd_cast.exe unit.api.compress.exe unit.core.count_true.exe unit.memory.store.exe unit.memory.store_equivalent.exe -j 3
```
[the remark this answers is not in the document]
Comment (jfalcou) [reply .github/workflows/unit.yml:150]: You can duplicate the section and have the non compress test run at 5 and the other at previous # of jobs

File: include/eve/module/core/regular/impl/simd/riscv/count_true.hpp (hunk 7 above)
Comment (ita-sc) [reply include/eve/module/core/regular/impl/simd/riscv/count_true.hpp:47]: You mean call there `eve::count_true[eve::ignore_none](...)`? Done

File: test/unit/api/compress/compress_test_filter.hpp (hunk 9 above)
Comment (DenisYaroshevskiy) [reply test/unit/api/compress/compress_test_filter.hpp:28]: maximum register of 16 bytes is not good enough. It means you never test any multiple bigger than 1. 

What is your problem? compiler running out of memory?

Comment (DenisYaroshevskiy) [reply test/unit/api/compress/compress_test_filter.hpp:28]: We already have a very large filtering on types.
Because not all of the interesting types are in the pool, it is very hard to judge what's tested. 

What if we remove just some of the signs?

Smth like:

```
return eve::detail::is_one_of<eve::element_type_t<T>>(
 types<std::int8_t, std::uint16_t, float, double>{}
);
```

is one of function is defined here: https://github.com/jfalcou/eve/blob/11c8d8c8ee2c9b42fa3a25613d6f5bc1efece991/include/eve/detail/meta.hpp#L75

I feel that in order to understand what you are doing here otherwise, you need to go into test and see what's removed there.

File: .github/workflows/unit.yml (hunk 10 above)
[the remark this answers is not in the document]
Comment (DenisYaroshevskiy) [reply .github/workflows/unit.yml]: This is definetly something to have a look at, if you have troubles with memory.

File: test/unit/api/compress/compress_test_filter.hpp (hunk 9 above)
Comment (ita-sc) [reply test/unit/api/compress/compress_test_filter.hpp:28]: You mean in compress filter try to specify exact types, instead of reducing fully? Yep, let's try. You suggest to check `std::int8_t, std::uint16_t, float, double` only?

File: .github/workflows/unit.yml (hunk 10 above)
[the remark this answers is not in the document]
Comment (ita-sc) [reply .github/workflows/unit.yml]: Yep, done

File: test/unit/api/compress/compress_test_filter.hpp (hunk 9 above)
Comment (DenisYaroshevskiy) [reply test/unit/api/compress/compress_test_filter.hpp:17]: > `exact types, instead of reducing fully`? 

What do you mean by reducing fully? I'm just looking for some filtering that make more sense.

Comment (ita-sc) [reply test/unit/api/compress/compress_test_filter.hpp:17]: Please take a look now. Did I understand you right?

File: test/tts/tts.hpp [hunk 11]
```diff
@@ -445,6 +445,21 @@ namespace tts
...
+  is_one_of(types<Ts...> const&) noexcept
+  {
+    bool found[] = {std::is_same_v<T, Ts>...};
+
+    return [&]<std::size_t... I>(std::index_sequence<I...> const&)
+    {
+      return (false || ... || found[I]);
+    }
+    (std::make_index_sequence<sizeof...(Ts)> {});
+  }
```
Comment (DenisYaroshevskiy) [inline test/tts/tts.hpp:461]: Yhis shouldn't be in tts

Just use https://github.com/jfalcou/eve/blob/11c8d8c8ee2c9b42fa3a25613d6f5bc1efece991/include/eve/detail/meta.hpp#L75

A;so you don't need the nested lambda

**Review by DenisYaroshevskiy — COMMENTED** [review]
Please remove your  is one of and gtg

File: test/tts/tts.hpp (hunk 11 above)
Comment (ita-sc) [reply test/tts/tts.hpp]: Thanks, fixed.

**Review by DenisYaroshevskiy — APPROVED** [review]
